src/bootstrap/lib.rs

```diff
@@ -447,17 +447,24 @@ impl Build {
         let bootstrap_out = if std::env::var("BOOTSTRAP_PYTHON").is_ok() {
             out.join("bootstrap").join("debug")
         } else {
-            let workspace_target_dir = std::env::var("CARGO_TARGET_DIR")
+            std::env::var("CARGO_TARGET_DIR")
                 .map(PathBuf::from)
-                .unwrap_or_else(|_| src.join("target"));
-            let bootstrap_out = workspace_target_dir.join("debug");
-            if !bootstrap_out.join("rustc").exists() {
-                // this restriction can be lifted whenever https://github.com/rust-lang/rfcs/pull/3028 is implemented
-                panic!("run `cargo build --bins` before `cargo run`")
-            }
-            bootstrap_out
+                .unwrap_or_else(|_| src.join("target"))
+                .join("debug")
         };
 
+        if bootstrap_out.join("bootstrap").exists() && !bootstrap_out.join("rustc").exists() {
+            let bin = bootstrap_out.join("bootstrap");
+            fs::hard_link(&bin, bootstrap_out.join("rustc"))
+                .expect("failed to link rustc to bootstrap shim");
```

Comment:
Build::copy defeats the point, it means the new files won't be updated on changes.

I don't know a good solution here :/ maybe we could do test somewhere of whether the fs supports file links and copy unconditionally if so?

Reply:
> it means the new files won't be updated on changes

Not sure what you meant. The current implementation is as follows:

```rust
    /// Copies a file from `src` to `dst`
    pub fn copy(&self, src: &Path, dst: &Path) {
        if self.config.dry_run {
            return;
        }
        self.verbose_than(1, &format!("Copy {:?} to {:?}", src, dst));
        if src == dst {
            return;
        }
        let _ = fs::remove_file(&dst);
        let metadata = t!(src.symlink_metadata());
        if metadata.file_type().is_symlink() {
            let link = t!(fs::read_link(src));
            t!(symlink_file(link, dst));
        } else if let Ok(()) = fs::hard_link(src, dst) {
            // Attempt to "easy copy" by creating a hard link
            // (symlinks don't work on windows), but if that fails
            // just fall back to a slow `copy` operation.
        } else {
            if let Err(e) = fs::copy(src, dst) {
                panic!("failed to copy `{}` to `{}`: {}", src.display(), dst.display(), e)
            }
            t!(fs::set_permissions(dst, metadata.permissions()));
            let atime = FileTime::from_last_access_time(&metadata);
            let mtime = FileTime::from_last_modification_time(&metadata);
            t!(filetime::set_file_times(dst, atime, mtime));
        }
    }
```

Reply:
No, because if you hit the fs::copy path, then the next time bootstrap is executed "rustc".exists() will return true and it will never recopy the binary. Which defeats the whole point of this change.

Reply:
That is true, I will just remove the check for `rustc.exists()` because it was triggering a filesystem error because the file already existed.

Reply:
Ah, I suppose that works too. The hard link should be very cheap compared to everything else bootstrap does 👍